File: app/src/main/java/com/ohmcoe/vultr/ServerListFragment.java
package com.ohmcoe.vultr;

import android.app.Fragment;
import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.Toast;

import java.io.IOException;

import okhttp3.OkHttpClient;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;


public class ServerListFragment extends Fragment{

    View serverListLayout;
    private ProgressDialog progressDialog;
    private ServerList serverList;
    private String APIKey;
    private ListView txtServerList;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        serverListLayout = inflater.inflate(R.layout.fragment_server_list, container, false);

        progressDialog = new ProgressDialog(getActivity());
        progressDialog.setTitle("Loading");
        progressDialog.setMessage("Wait while loading...");
        progressDialog.setCancelable(false);

        Bundle bundle = getArguments();
        APIKey = bundle.getString("API-Key");


        prepareServerList();


        return serverListLayout;
    }

    protected void prepareServerList()
    {
        serverList = new ServerList();
        txtServerList = (ListView)serverListLayout.findViewById(R.id.txtServerList);
        getServerList();
    }

    protected void updateUI()
    {
        ServerAdapter serverAdapter = new ServerAdapter(getContext(), R.layout.server_list, serverList.toList());
        txtServerList.setAdapter(serverAdapter);
        txtServerList.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                Server server = (Server)parent.getItemAtPosition(position);
                Bundle bundle = new Bundle();
                bundle.putString("API-Key", APIKey);
                bundle.putString("SUBID", server.getSUBID());
                Fragment serverFragment = new ServerFragment();
                serverFragment.setArguments(bundle);
                getFragmentManager().beginTransaction()
                        .replace(R.id.content_frame, serverFragment)
                        .addToBackStack(null)
                        .commit();
            }
        });
    }

    protected void getServerList() {
        progressDialog.show();

        String API_BASE_URL = getResources().getString(R.string.base_uri);

        OkHttpClient.Builder httpClient = new OkHttpClient.Builder();

        Retrofit.Builder builder = new Retrofit.Builder()
                .baseUrl(API_BASE_URL)
                .addConverterFactory(GsonConverterFactory.create());

        Retrofit retrofit = builder
                .client(httpClient.build())
                .build();

        VultrClient client = retrofit.create(VultrClient.class);
        Call<ResponseBody> call = client.getServerList(APIKey);

        call.enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                try {
                    if (response.code() == 200) {
                        String body = response.body().string();
                        serverList = new ServerList(body);
                        updateUI();
                    }
                    else
                    {
                        CharSequence text = "Loading Failure response code " + response.code();
                        Toast toast = Toast.makeText(getActivity(), text, Toast.LENGTH_LONG);
                        toast.show();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }

                progressDialog.dismiss();
            }

            @Override
            public void onFailure(Call<ResponseBody> call, Throwable t) {
                progressDialog.dismiss();
            }
        });
    }
}

File: app/src/main/java/com/ohmcoe/vultr/ServerListFragment.kt
package com.ohmcoe.vultr

import android.app.Fragment
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ListView
import kotlinx.android.synthetic.main.fragment_server_list.view.*
import okhttp3.OkHttpClient
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.IOException

class ServerListFragment : Fragment() {

    private lateinit var waitDialog: WaitDialog
    private lateinit var serverListLayout: View
    private lateinit var APIKey: String
    private lateinit var txtServerList: ListView
    private var serverList: ServerList? = null


    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View {
        serverListLayout = inflater!!.inflate(R.layout.fragment_server_list, container, false)

        waitDialog = WaitDialog.newInstance()

        val bundle = arguments
        APIKey = bundle.getString("API-Key")

        serverList = ServerList()
        txtServerList = serverListLayout.txtServerList


        //restore instance
        if (savedInstanceState == null) {
            getServerList()
        } else {
            serverList = savedInstanceState.getParcelable("serverList")
            if (serverList == null)
                getServerList()
            else
                updateUI()
        }

        return serverListLayout
    }

    private fun dismissDialog() {
        waitDialog.dismiss()
    }

    private fun showDialog() {
        childFragmentManager.executePendingTransactions()
        waitDialog.isCancelable = false
        if (!waitDialog.isAdded)
            waitDialog.show(childFragmentManager, "waitDialog")
    }


    private fun updateUI() {
        if (activity == null || serverList == null)
            return;

        val serverAdapter = ServerAdapter(activity, R.layout.server_list, serverList!!.toList()!!)
        txtServerList.adapter = serverAdapter
        txtServerList.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id ->
            val server = parent.getItemAtPosition(position) as Server
            val bundle = Bundle()
            bundle.putString("API-Key", APIKey)
            bundle.putString("SUBID", server.subid)
            val serverFragment = ServerFragment()
            serverFragment.arguments = bundle
            fragmentManager.beginTransaction()
                    .replace(R.id.content_frame, serverFragment)
                    .addToBackStack(null)
                    .commit()
        }
    }


    private fun getServerList() {
        showDialog()

        val API_BASE_URL = resources.getString(R.string.base_uri)

        val httpClient = OkHttpClient.Builder()

        val builder = Retrofit.Builder()
                .baseUrl(API_BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())

        val retrofit = builder
                .client(httpClient.build())
                .build()

        val client = retrofit.create(VultrClient::class.java)
        val clientCall = client.getServerList(APIKey)

        clientCall.enqueue(object : Callback<ResponseBody> {
            override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
                try {
                    if (response.code() == 200) {
                        val body = response.body()!!.string()
                        serverList = ServerList(body)
                        updateUI()
                    } else {
                        val text = "Loading Failure response code " + response.code()
                        val myToast = MyToast(activity, text)
                        myToast.show()
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                }

                dismissDialog()
            }

            override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
                dismissDialog()
            }
        })
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)

        outState?.putParcelable("serverList", serverList)
    }
}// Required empty public constructor
